### PR TITLE
Add option to repeat latched messages at the start of bag splits

### PR DESCRIPTION
**Pull request by tappan-at-git**
Split bagfiles lose all latched data. Many long-running ROS applications will include some kind of latched data that's only published once on launch (e.g. maps, static transforms), but prefer split bagfiles for ease of use, meaning they will have a difficult time actually using latched data without playing back the entire series of bagfiles.

This PR adds the --repeat-latched option to `rosbag record`. When this option is enabled, `rosbag::Recorder` remembers up to one message for each <topic, publisher> pair and prepends each latched message to each new bag file.

### CODE REVIEW

**Comment by mikepurvis** [comment]
This looks like a great addition; would you consider adding a simple test for the functionality, however? See several other bits of latch-related functional tests here:

https://github.com/ros/ros_comm/tree/melodic-devel/test/test_rosbag/test

**Comment by dirk-thomas** [comment]
@tappan-at-git Friendly ping.

File: tools/rosbag/src/rosbag/rosbag_main.py [hunk 1]
```diff
@@ -368,7 +370,7 @@ def eval_fn(topic, m, t):
                 else:
                     print('NO MATCH', verbose_pattern(topic, msg, t))          
 
-                total_bytes += len(serialized_bytes) 
+                total_bytes += len(serialized_bytes)
```
Comment (dirk-thomas) [inline tools/rosbag/src/rosbag/rosbag_main.py]: Please revert unrelated whitespace changes.

Same below.

**Comment by dirk-thomas** [comment]
Since this patch changes ABI I will mark it to be considered for a future distro only.

File: tools/rosbag/src/rosbag/rosbag_main.py (hunk 1 above)
Comment (tappan-at-git) [reply tools/rosbag/src/rosbag/rosbag_main.py]: I really thought I caught all of the EOL changes my editor snuck in. Will address along with the requested tests when I have the cycles.

**Comment by oceanusxiv** [comment]
This resolves #706 if I recall correctly.

File: tools/rosbag/src/rosbag/rosbag_main.py (hunk 1 above)
Comment (dirk-thomas) [reply tools/rosbag/src/rosbag/rosbag_main.py]: @tappan-at-git Please follow up on this soon if you want this patch to be considered for Noetic.

Comment (dirk-thomas) [reply tools/rosbag/src/rosbag/rosbag_main.py]: > Please follow up on this soon if you want this patch to be considered for Noetic.

@paulbovbel FYI

Comment (dirk-thomas) [reply tools/rosbag/src/rosbag/rosbag_main.py]: @tappan-at-git Friendly ping.

**Comment by tappan-at-git** [comment]
Sorry about the force-push spam. Github was still showing one of the whitespace changes after I had pushed a fix. Unless Github is _seriously_ messing with me, all the unrelated whitespace changes should be squashed at this point.

I can't commit to adding full testing for this feature. Unless I'm missing something [rosbag's existing latched message test](https://github.com/ros/ros_comm/blob/noetic-devel/test/test_rosbag/test/latched_pub.py) doesn't seem to check if its latched messages were recorded correctly, and I don't see any tests for splits. I just don't have time to add those.

**Comment by dirk-thomas** [comment]
Thank you for the patch.

It is unfortunate that it doesn't come with any kind of test coverage. The fact that some other parts of existing code aren't covered is not a reason to not implement tests for new features.

Since it is ABI breaking I will go ahead and merge it for the upcoming Noetic release anyway sice otherwise this feature couldn't land afterwards.

**Comment by dirk-thomas** [comment]
@ros-pull-request-builder retest this please
